grails-app/services/org/pih/warehouse/order/OrderService.groovy

```diff
@@ -869,7 +869,19 @@ class OrderService {
                     if (currentLocation.isAccountingRequired() && !code) {
                         throw new IllegalArgumentException("Budget code is required.")
                     }
-                    BudgetCode budgetCode = BudgetCode.findByCode(code)
+
+                    // There can be more than one budget code with the same code, despite the fact that the code is unique from the domain perspective.
+                    // As a fix for that we would like to have only one active budget code with the same code, so we have to find only active
+                    // budget codes. In case when there is only one budget code, and this one is inactive we have to throw a validation error,
+                    // so we can't just look for BudgetCode.findAllByCodeAndActive(code, true);
+                    List<BudgetCode> foundBudgetCodes = BudgetCode.findAllByCode(code)
+                    List<BudgetCode> activeBudgetCodes = foundBudgetCodes.findAll { it.active }
+
+                    if (activeBudgetCodes.size() > 1) {
+                        throw new IllegalArgumentException("Found more than one active budget code with the same code.")
+                    }
+
+                    BudgetCode budgetCode = activeBudgetCodes.size() == 0 ? foundBudgetCodes[0] : activeBudgetCodes.first()
```

Comment:
> you're mixing your access methods here ([0] and x.first())

I did it on purpose, the `.first` method throws an error when there is no element, `[0]` doesn't so it will be null, and then we will get the information that we didn't find any budget code.

If there is more than one inactive budget code, we will get the error that the budget code is inactive, independently of which one inactive budget code the system will pick.

Reply:
oh neat I didn't actually know that groovy made index-based accesses null safe. :+1: